(CDAP-16353) Support preview to run with non-sharing local level DB.

wyzhang opened the pull request:
Why:
    Allow preview to run with its own local levelDB. This allows us to
    deploy cdap in a distributed fashion without shared underlying
    storage when storage impl is NOSQL.

    What:
    This change allow preview to run with local levelDB when data storage
    is NOSQL. It achieves this by binding interface to local vs remote
    implementation dynamically depending on storage implementation config.

    The crux of the change is in DefaultPreviewRunnerModule where we bind
    following interfaces to local impl for SQL and remote for NOSQL
    * ArtifactRepositoryReader
    * PluginFinder
    * PreferencesFetcher

    Changes in tests:
    * Starting AppFabric in TestBase (required by PreviewDataPipelineTest) as
      we use NOSQL in unit tests, therefore bindings use remote impls for
      above interfaces and need to talk to AppFabric
    * PreviewServiceMainTest now no longer needs to workaround the issue of
      non-sharing local levelDB (previously it needs to shutdown AppFabric
      before starting preview so as to have levelDB state updated by AppFabric
      to be visible to preview in order for it to work)
    * Extend PreviewServiceMainTest with a new test case where app has plugin
      (in order to cover plugin finder code path in preview), thus better
      test coverage.

albertshau commented on cdap-master/src/test/java/io/cdap/cdap/master/environment/k8s/PreviewServiceMainTest.java:
```diff
@@ -39,82 +42,223 @@
+    // from appFabric when running a preview
+    startService(PreviewServiceMain.class);
 
-    // have to stop AppFabric so that Preview can share the same leveldb table
```
nice that we don't need this anymore

wyzhang replied:
Ideally we should have another test that spin up SQL. If time permits, will look into add that.

albertshau commented on cdap-unit-test/src/main/java/io/cdap/cdap/test/TestBase.java:
```diff
@@ -369,6 +371,8 @@ protected void configure() {
     provisioningService.startAndWait();
     metadataSubscriberService.startAndWait();
+    appFabricServer = injector.getInstance(AppFabricServer.class);
+    appFabricServer.startAndWait();
```
not really sure how things were working before if app fabric was never getting started before. Unless I'm missing something, it doesn't seem like the changes in this PR would affect TestBase?

wyzhang replied:
In unit tests, we always use NOSQL thus will always exercise the remote call code path. 

Removed for now.

Thought some other preview tests are failing due to AppFabric being not up. Let me run all tests again and see if anything is broken without this.

wyzhang replied:
@albertshau Without starting AppFabric, PreviewDataPipelineTest fails as we use NOSQL in test and preview in this case fetch artifact, preferences from appfabric. Previously preview just access local levelDB directly. Need to bring back this starting up appfabric.

wyzhang commented:
Without starting AppFabric, PreviewDataPipelineTest fails. Need to bring back appfabric in TestBase.

wyzhang committed:
(CDAP-16353) Support preview to run with non-sharing local level DB.

Why:
Allow preview to run with its own local levelDB. This allows us to
deploy cdap in a dsitributed fashion without shared underlyihg
storage when storage impl is NOSQL.

What:
This change allow preview to run with local levelDB when data storage
is NOSQL. It achieves this by binding interface to local vs remote
implementation dynamically depending on storage impelementation config.

The crux of the chagne is in DefaultPreviewRunnerModule where we bind
following interafces to local impl for SQL and remote for NOSQL
* ArtifactRepositoryReader
* PluginFinder
* PreferencesFetcher

Changes in tests:
* Starting AppFabric in TestBase (required by PreviewDataPipelineTest) as
  we use NOSQL in unit tests, therefore bindings use remote impls for
  above interafces and need to talk to AppFabric
* PreviewServiceMainTest now no longer needs to workaround the issue of
  non-sharing local levelDB (previously it needs to shutdown AppFabric
  before starting preview so as to have levelDB state updated by AppFabric
  to be visible to preview in order for it to work)
* Extend PreviewServiceMainTest with a new test case where app has plugin
  (in order to cover plugin finder code path in preview), thus better
  test coverage.
